dali/kernels/imgproc/convolution/laplacian_cpu.h

```diff
@@ -0,0 +1,299 @@
+ * If it is OpenCV style convolution that applies smoothing in perpendicular dimensions,
+ * SeparableConvolution kernel is used to compute multidimensional convolution,
+ * otherwise (if window sizes in non derivative dimensions are one) skips unnecessary
+ * convolutions, performing only single one dimensional convolution with derivative kernel.
+ */
+template <typename Out, typename In, typename W, int axes, int deriv_axis,
+          bool has_channels = false, typename T = conv_transform::TransScaleSat<Out, W>>
+struct Convolution {
+  using MultiDimConv = SeparableConvolutionCpu<Out, In, W, axes, has_channels, T>;
+  static constexpr int ndim = MultiDimConv::ndim;
+  using SingleDimConv = ConvolutionCpu<Out, In, W, ndim, deriv_axis, has_channels, T>;
+
+  bool IsNoSmooting(const std::array<int, axes>& window_sizes) {
+    for (int axis = 0; axis < axes; axis++) {
+      if (axis != deriv_axis && window_sizes[axis] != 1) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  KernelRequirements Setup(KernelContext& ctx, const TensorShape<ndim>& in_shape,
+                           const std::array<int, axes>& window_sizes) {
+    no_smoothing_ = IsNoSmooting(window_sizes);
+    if (no_smoothing_) {
+      if (!single_dim_impl_) {
+        multi_dim_impl_.reset();
+        single_dim_impl_ = std::make_unique<SingleDimConv>();
+      }
+      return single_dim_impl_->Setup(ctx, in_shape, window_sizes[deriv_axis]);
+    }
+    if (!multi_dim_impl_) {
+      single_dim_impl_.reset();
+      multi_dim_impl_ = std::make_unique<MultiDimConv>();
+    }
+    return multi_dim_impl_->Setup(ctx, in_shape, window_sizes);
+  }
+
+  void Run(KernelContext& ctx, const TensorView<StorageCPU, Out, ndim> &out,
+           const TensorView<StorageCPU, const In, ndim>& in,
+           const std::array<TensorView<StorageCPU, const W, 1>, axes>& windows,
+           const T& transform = {}) {
+    if (no_smoothing_) {
+      single_dim_impl_->Run(ctx, out, in, windows[deriv_axis], transform);
+    } else {
+      multi_dim_impl_->Run(ctx, out, in, windows, transform);
+    }
+  }
+
+  bool no_smoothing_;
+  std::unique_ptr<MultiDimConv> multi_dim_impl_;
+  std::unique_ptr<SingleDimConv> single_dim_impl_;
+};
+
+
+/**
+ * @brief Provides Laplacian specializations for 2 and 3 dimensional data.
+ *  Run methods expect additional ``acc`` buffer that will be used to accumulate
+ *  partial derivatives. If ``Intermediate`` and ``Out`` are the same type,
+ *  the `acc` and ``out`` can be the same tensor.
+ */
+template <typename T, typename Intermediate, typename Out, typename In, typename W, int axes,
+          bool has_channels>
+struct LaplacianCPUBase;
+
+template <typename T, typename Intermediate, typename Out, typename In, typename W,
+          bool has_channels>
+struct LaplacianCPUBase<T, Intermediate, Out, In, W, 2, has_channels> {
+  static constexpr int axes = 2;
+  using DxKernel = Convolution<Intermediate, In, W, axes, 0, has_channels,
```

Comment:
As mentioned above, the convolution in 2D case, assumes the HW[C] layout, so the x is typically the second (index 1) data axis.

Reply:
I reversed the x, y, z naming of the sub kernels.